homework/homework.js

```diff
@@ -0,0 +1,181 @@
+'use strict';
+
+class ShopError extends Error {
+    
+    static ERROR_CODE = {
+        EXPIRED_ALIMENT: 'EXPIRED_ALIMENT',
+        NOT_ENOUGH_FUNDS: 'NOT_ENOUGH_FUNDS'
+    };
+
+    static ERROR_MESSAGE = {
+        [ShopError.ERROR_CODE.EXPIRED_ALIMENT]: 'Customer can not add an expired aliment to their shopping cart.',
+        [ShopError.ERROR_CODE.NOT_ENOUGH_FUNDS]: 'Customer does not have enough funds to pay for all their products.'
+    };
+
+    constructor({ code, message = ShopError.ERROR_MESSAGE[code], data}) {
+        super(message);
+
+        this.code = code;
+        this.data = data;
+    }
+
+    toString() {
+
+        return JSON.stringify({
+            code: this.code,
+            message: this.message,
+            data: this.data
+        });
+    }
+
+}
+
+
+class Person {
+    constructor(name) {
+        this.name = name;
+    }
+
+}
+
+class Customer extends Person {
+
+    constructor(name, money) {
+        super(name);
+        this.money = money;
+        this.shoppingCart = [];
+    }
+
+    addToCart(product) {
+        let currentDate = new Date();
+        if (product.expirationDate < currentDate) {
+            throw new ShopError({
+                code: ShopError.ERROR_CODE.EXPIRED_ALIMENT,
+                data: {
+                    expirationDate: product.expirationDate,
+                    currentDate: currentDate
+                }
+            });
+        }
+
+        this.shoppingCart.push(product);
+    }
+
+    removeFromCart(product) {
+        let indexOfProduct = this.shoppingCart.indexOf(product);
+
+        if (index != -1) {
+            this.shoppingCart.splice(indexOfProduct, 1);
+        }
+    }
+
+}
+
+class Cashier extends Person {
+
+    constructor(name, totalSales) {
+        super(name);
+        this.totalSales = totalSales;
+    }
+
+    scanAndCash(customer) {
+        let sumToPay = 0;
+        
+        for (let product of customer.shoppingCart) {
+            sumToPay += product.price;
+        }
+
+        if (sumToPay > customer.money) {
+
+            throw new ShopError({
+                code: ShopError.ERROR_CODE.NOT_ENOUGH_FUNDS,
+                data: {
+                    customerName: customer.name,
+                    funds: customer.money,
+                    needToPay: sumToPay
+                }
+            });
+        }
+
+        customer.shoppingCart = [];
```

Comment:
`shoppingCart` and `money` properties should be private and you should only expose them via getters as read-only properties, otherwise any other entity can play with `Customer` internals.

If you want to manipulate them you should implement methods in the `Customer` class and call them where needed.